### ISSUE TITLE
False positive on `UnreachableCode`

### ISSUE THREAD

**Issue by eygraber** [detekt/detekt#4676]
Using detekt 1.20.0-RC2

The following code results in 2 `UnreachableCode` violations on lines 2 and 4, but I don't see how those are unreachable:

```
private suspend fun transitionToState(mode: Tab) {
    val (arrow, search) = searchBoxDrawables

    when(binding.root.currentState) {
      mode.stateId -> mode.fraction.let { fraction ->
        search.alpha = fraction.searchInterpolation
        arrow.alpha = fraction.arrowInterpolation
      }

      else -> suspendCancellableCoroutine { continuation ->
        val listener = object : TransitionAdapter() {
          override fun onTransitionChange(motionLayout: MotionLayout?, startId: Int, endId: Int, progress: Float) {
            search.alpha = progress.searchInterpolation
            arrow.alpha = progress.arrowInterpolation
          }

          override fun onTransitionCompleted(motionLayout: MotionLayout?, currentId: Int) {
            binding.root.removeTransitionListener(this)
            continuation.resume(Unit)
          }
        }

        continuation.invokeOnCancellation { binding.root.removeTransitionListener(listener) }
        binding.root.addTransitionListener(listener)
        binding.root.transitionToState(mode.stateId)
      }
    }
  }
```

**Comment by eygraber** [comment]
Same issue with lines 4, 11, 16, 27, and 34:

```
override suspend fun findMapPinCluster(latLng: Point): MapPinCluster? {
    val map = binding.mapView.getMapboxMap()

    val cluster =
      binding
        .mapView
        .findFeatures(latLng, Map.Layers.MapPinClusters.id)
        .firstOrNull { feature -> feature.feature.getBooleanProperty("cluster") == true }
        ?: return null

    val zoom = binding.mapView.findGeoJsonClusterExpansionZoom(
      id = Map.Sources.MapPins.id,
      cluster = cluster.feature
    ) ?: return null

    val features = if(map.cameraState.zoom == map.getBounds().maxZoom) {
      binding.mapView.findGeoJsonClusterLeaves(
        id = Map.Sources.MapPins.id,
        cluster = cluster.feature,
        limit = cluster.feature.getNumberProperty("point_count").toLong()
      ) ?: listOf(cluster.feature)
    }
    else {
      listOf(cluster.feature)
    }

    val point =
      cluster
        .feature
        .geometry()
        ?.cast<Point>()
        ?: return null

    return MapPinCluster(
      features = features,
      zoom = LatLngZoom(
        lat = point.latitude(),
        lng = point.longitude(),
        zoom = zoom.toDouble()
      )
    )
  }
```

**Comment by cortinico** [comment]
Is this a regression?

**Comment by eygraber** [comment]
This issue, as well as the other ones I filed today, are not present in 1.19.

I'm not sure type resolution was working fully in 1.19 and earlier because it finds no violations in my project. After updating to 1.20.0-RC1/2 I get hundreds of violations. 

Some of them are false positives (i.e. the ones that I filed issues for), but a large majority of them are actually violations (i.e. false negatives in 1.19 and earlier.

**Comment by BraisGabin** [comment]
🤔 That's odd, I can't reproduce it on my project... But it seems something related with the compiler doesn't know about what `binding` is.

**Comment by eygraber** [comment]
Is it possible that something like https://github.com/detekt/detekt/pull/4684 is needed to fix it?

**Comment by BraisGabin** [comment]
It seems so.

**Comment by kvn-stgl** [comment]
Well, the test code for this is really short:
```kotlin
        @Test
        fun `test`() {
            val code = """
                fun test(): Boolean {
                    binding.undeclaredVal ?: return false
                    return true
                }
            """

            assertThat(subject.lintWithContext(env, code)).isEmpty()
        }
```

But I really don't know how we can fix it. For the compiler it seems like `binding.undeclaredVal` is null (because it's undeclared) and with this assumption the `return true`-part is really unreachable.

**Comment by eygraber** [comment]
Why does the compiler think that it's undeclared though? The class and field are both present in the generated code.

**Comment by BraisGabin** [comment]
> Why does the compiler think that it's undeclared though? The class and field are both present in the generated code.

That's a known issue/known limitation that detekt have right now. We need to move to a kotlin plugin to fix that. We have something here: https://github.com/detekt/detekt-compiler-plugin but in kotlin 1.7/1.8 they will break all of that so we are waiting for that breaking change and the api stability to implement it properly. But I must say that you can use that repo, it works but the edges are a bit sharp.

**Comment by eygraber** [comment]
So what's the proper solution in the interim, considering compiler plugins probably won't be stable for another year or two? Something was definitely broken with type resolution pre 1.20.0, so I don't want to be stuck there.

I tried using the `detekt-compiler-plugin` but ran into a bunch of issues (probably because it's an old build). I can always suppress these rules because they're not showstoppers for me, but if this issue grows to affect other rules that I rely on I'd be stuck.

**Comment by BraisGabin** [comment]
@eygraber could you check the detekt `SNAPSHOT` to see if you can reproduce the same issue after we merge #4723?

**Comment by eygraber** [comment]
The issue is resolved in `SNAPSHOT`
